components/gitpod-cli/cmd/rebuild.go

```diff
@@ -0,0 +1,241 @@
+	containerIds, err := cmd.Output()
+	if err != nil {
+		return err
+	}
+
+	for _, id := range strings.Split(string(containerIds), "\n") {
+		if len(id) == 0 {
+			continue
+		}
+
+		cmd = exec.Command("docker", "stop", id)
+		err := cmd.Run()
+		if err != nil {
+			return err
+		}
+
+		cmd = exec.Command("docker", "rm", "-f", id)
+		err = cmd.Run()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func runRebuild(ctx context.Context, supervisorClient *supervisor.SupervisorClient, event *utils.EventTracker) error {
+	wsInfo, err := supervisorClient.Info.WorkspaceInfo(ctx, &api.WorkspaceInfoRequest{})
+	if err != nil {
+		event.Set("ErrorCode", utils.SystemErrorCode)
+		return err
+	}
+
+	tmpDir, err := os.MkdirTemp("", "gp-rebuild-*")
+	if err != nil {
+		event.Set("ErrorCode", utils.SystemErrorCode)
+		return err
+	}
+	defer os.RemoveAll(tmpDir)
+
+	gitpodConfig, err := utils.ParseGitpodConfig(wsInfo.CheckoutLocation)
+	if err != nil {
+		fmt.Println("The .gitpod.yml file cannot be parsed: please check the file and try again")
+		fmt.Println("")
+		fmt.Println("For help check out the reference page:")
+		fmt.Println("https://www.gitpod.io/docs/references/gitpod-yml#gitpodyml")
+		event.Set("ErrorCode", utils.RebuildErrorCode_MalformedGitpodYaml)
+		return err
+	}
+
+	if gitpodConfig == nil {
+		fmt.Println("To test the image build, you need to configure your project with a .gitpod.yml file")
+		fmt.Println("")
+		fmt.Println("For a quick start, try running:\n$ gp init -i")
+		fmt.Println("")
+		fmt.Println("Alternatively, check out the following docs for getting started configuring your project")
+		fmt.Println("https://www.gitpod.io/docs/configure#configure-gitpod")
+		event.Set("ErrorCode", utils.RebuildErrorCode_MissingGitpodYaml)
+		return err
+	}
+
+	var baseimage string
+	switch img := gitpodConfig.Image.(type) {
+	case nil:
+		baseimage = ""
+	case string:
+		baseimage = "FROM " + img
+	case map[interface{}]interface{}:
+		dockerfilePath := filepath.Join(wsInfo.CheckoutLocation, img["file"].(string))
+
+		if _, err := os.Stat(dockerfilePath); os.IsNotExist(err) {
+			fmt.Println("Your .gitpod.yml points to a Dockerfile that doesn't exist: " + dockerfilePath)
+			event.Set("ErrorCode", utils.RebuildErrorCode_DockerfileNotFound).Send(ctx)
+			return err
+		}
+		dockerfile, err := os.ReadFile(dockerfilePath)
+		if err != nil {
+			event.Set("ErrorCode", utils.RebuildErrorCode_DockerfileCannotRead)
+			return err
+		}
+		if string(dockerfile) == "" {
+			fmt.Println("Your Gitpod's Dockerfile is empty")
+			fmt.Println("")
+			fmt.Println("To learn how to customize your workspace, check out the following docs:")
+			fmt.Println("https://www.gitpod.io/docs/configure/workspaces/workspace-image#use-a-custom-dockerfile")
+			fmt.Println("")
+			fmt.Println("Once you configure your Dockerfile, re-run this command to validate your changes")
+			event.Set("ErrorCode", utils.RebuildErrorCode_DockerfileEmpty)
+			return err
```

Comment:
@andreafalzetti @felladrin for me this branch is hanging 🤔 

I think the problem is that it returns nil here. It should return some real error. We need to double check all places, where we return without having err object in advance.

Also adding os.Exit(0) in the top level routine at the end makes sense to make sure to exit.

Reply:
i.e. something like:
```go
ok, err = runRebuild(ctx, supervisorClient, event)
if event.Data.ErrorCode == "" {
	if err != nil {
		event.Set("ErrorCode", utils.SystemErrorCode)
	} else if !ok {
		event.Set("ErrorCode", utils.UserErrorCode)
	}
}
event.Send(ctx)

if err != nil {
	utils.LogError(ctx, err, "Failed to rebuild", supervisorClient)
}
var exitCode int
if err != nil || !ok {
	exitCode = 1
}
os.Exit(exitCode)
```

Reply:
oops, I wonder how I missed this :(

will send a follow-up PR!

Reply:
there is also the default branch of the switch having the same issue, I will check them all